src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,175 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node newNode = new Node<>(value);
+        if (size == 0) {
+            head = tail = newNode;
+            size++;
+            return;
+        }
+        newNode.previous = tail;
+        tail.next = newNode;
+        tail = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == 0) {
+            addAsHead(value);
+            return;
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+        checkIndex(index);
+        Node newNode = new Node(value);
+        Node currentNode = getNodeByIndex(index);
+        insertBefore(currentNode, newNode);
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        T replacedValue;
+        checkIndex(index);
+        Node current = getNodeByIndex(index);
+        replacedValue = (T) current.value;
+        current.value = value;
+        return replacedValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        return unlinkNode(index);
     }
 
     @Override
     public boolean remove(T object) {
+        Node currentNode = head;
+        for (int i = 0; i < size; i++) {
+            if (currentNode.value == object
+                    || currentNode.value.equals(object) && currentNode.value != null) {
+                unlinkNode(i);
```

Comment:
```suggestion
                unlinkNode(currentNode);
```
Here you are iterating over the whole list, you find the node!, then pass an index, and iterate again in the method uninkNode(). Basically doing the same thing twice